## Домашняя работа №1

In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv("data/weatherHistory.csv")
data.rename({'Formatted Date':'date', 'Temperature (C)':'temp', 'Apparent Temperature (C)':'temp_app',
                    'Precip Type':'weather_type', 'Wind Speed (km/h)':'wind_speed', 'Humidity':'humidity',
                    'Daily Summary':'summary'}, 
                    axis=1, inplace=True)
data.date = pd.to_datetime(data.date, utc=True)

1) самый холодный и самый ветренный дни в Сегеде за всю имеющуюся историю

In [34]:
data[data.temp == data.temp.min()]

date Summary weather_type       temp   temp_app  \
54847 2012-02-10 06:00:00+00:00   Foggy         snow -21.822222 -21.822222   

       humidity  wind_speed  Wind Bearing (degrees)  Visibility (km)  \
54847       0.8      3.0751                   323.0           1.3685   

       Loud Cover  Pressure (millibars)  \
54847         0.0               1033.66   

                                                 summary  
54847  Foggy starting in the morning continuing until...

In [35]:
data[data.wind_speed == data.wind_speed.max()]

date                              Summary  \
12181 2007-01-29 12:00:00+00:00  Dangerously Windy and Partly Cloudy   

      weather_type      temp  temp_app  humidity  wind_speed  \
12181         rain  8.944444  3.483333      0.49     63.8526   

       Wind Bearing (degrees)  Visibility (km)  Loud Cover  \
12181                   307.0          11.4471         0.0   

       Pressure (millibars)                                            summary  
12181               1009.05  Mostly cloudy throughout the day and windy sta...

2) сколько раз шел дождь летом 2014 года (все лето)

In [47]:
data[(data['date'] > '2014-06-1') & (data['date'] < '2014-09-01')].sort_values(by="date")\
                                                                  .weather_type\
                                                                  .value_counts()

rain    2204
Name: weather_type, dtype: int64

3) Самый ветрянный месяц

In [75]:
months = data[["date", "wind_speed"]].groupby(lambda idx: data.loc[idx].date.month_name())

In [80]:
months.aggregate(np.mean).sort_values(by="wind_speed")

wind_speed
August       8.942707
September    9.610204
June         9.632453
July         9.637124
October     10.008495
November    10.947953
May         10.953186
December    11.094578
January     11.515865
April       11.876139
February    12.180815
March       13.415825

В среднем, за всю историю самым ветренным является март.

Самый влажный год

In [81]:
years = data[["date", "humidity"]].groupby(lambda idx: data.loc[idx].date.year)

In [82]:
years.aggregate(np.mean).sort_values(by="humidity")

humidity
2012  0.689513
2007  0.689647
2008  0.701239
2009  0.707247
2015  0.732354
2011  0.736008
2014  0.748571
2013  0.754197
2016  0.760868
2006  0.767349
2010  0.796861
2005  0.890000

Самый влажный год: 2005

Неделя с самой переменчивой погодой

In [88]:
weeks = data[["date", "temp", "humidity", "wind_speed"]].groupby(lambda idx: data.loc[idx].date.weekofyear)

In [90]:
weeks.aggregate(np.std).sort_values(by=["temp", "humidity", "wind_speed"])

temp  humidity  wind_speed
48  3.617409  0.105251    6.351566
47  4.133849  0.108823    6.485779
49  4.155522  0.103713    7.918307
51  4.211092  0.084308    6.394904
50  4.280489  0.100274    7.077564
3   4.532339  0.107841    6.884513
7   4.628565  0.127673    7.037540
46  4.762689  0.120901    6.068201
2   4.811115  0.113069    6.531584
9   4.839419  0.153718    7.207053
43  4.911441  0.140614    6.717006
21  4.923256  0.188684    6.277847
10  4.947341  0.157593    8.092508
45  4.951843  0.128664    7.402578
8   4.967552  0.148057    7.138245
52  4.982740  0.108186    7.435814
1   5.008042  0.101536    7.393594
16  5.020456  0.208602    6.792677
19  5.025860  0.199540    7.063987
24  5.158063  0.203184    5.791146
41  5.161809  0.174645    5.479907
37  5.163760  0.209729    6.085962
53  5.182074  0.115782    6.664950
22  5.221413  0.183340    5.478476
44  5.231834  0.144379    7.299589
42  5.234595  0.167927    7.430465
5   5.251543  0.132209    8.692668
20  5.262296  0.184618    7.416471
11  5.265073  0.191548    8.342104
15  5.270850  0.197066    7.490980
14  5.272616  0.193015    8.070549
39  5.315730  0.194418    6.206881
13  5.334651  0.197237    8.477027
31  5.383029  0.210266    5.263082
17  5.413772  0.204314    6.125820
12  5.452593  0.190884    8.363720
4   5.482606  0.118193    7.777317
40  5.486404  0.183379    6.465498
18  5.522132  0.205129    6.891334
26  5.538135  0.207666    5.843644
30  5.587042  0.221947    6.041617
23  5.599345  0.194851    5.596402
27  5.677395  0.216169    6.027635
38  5.689703  0.188260    6.303168
32  5.709605  0.212906    5.425355
28  5.755464  0.205628    5.776108
25  5.794001  0.204837    5.872436
29  5.864455  0.216166    6.088937
35  5.875341  0.221038    6.228777
33  5.925675  0.216900    5.591991
36  6.043471  0.216281    5.979321
34  6.149509  0.223476    6.005965
6   6.456311  0.129590    7.094790

Статистически, 6-я неделя года хвастается самой переменчивой погодой

## Задание 2

In [93]:
data = pd.read_csv('data/train.csv')
data.dropna(subset=['description'], inplace=True) # так выкидываются строчки с пропусками в колонке description

In [94]:
data.head()

parent_category_name           category_name        param_1         param_2  \
0  Бытовая электроника                Телефоны         Xiaomi             NaN   
1            Транспорт              Автомобили     С пробегом            Audi   
2          Личные вещи  Детская одежда и обувь  Для мальчиков  Верхняя одежда   
3  Бытовая электроника                Телефоны         iPhone             NaN   
4             Животные                Аквариум            NaN             NaN   

                param_3                                              title  \
0                   NaN                                    Xiaomi RedMI 4a   
1                    Q5                                      Audi Q5, 2009   
2  98-104 см (2-4 года)                             Толстовка очень теплая   
3                   NaN  Продаю айфон 6 16 гб сильвер с зарядкой,без ко...   
4                   NaN                                           Аквариум   

                                         description     price  \
0  Телефон новый, не использованный, в заводской ...    6900.0   
1  Продам автомобиль AUDI SUV Stronic Quattro 2.0...  890000.0   
2  на меху, очень теплая, хорошее состояние , бол...     500.0   
3  Продаю айфон 6 16 гб сильвер с зарядкой,без ко...   23000.0   
4  Аквариум для маленьких рыбок+корм+грунт+подста...    1000.0   

   item_seq_number activation_date user_type  image_top_1  deal_probability  
0               22      2017-03-26   Private       2944.0           0.16933  
1                6      2017-03-16   Private       1132.0           0.26029  
2              123      2017-03-27   Private         61.0           0.80323  
3                7      2017-03-15   Private       2918.0           0.03486  
4              109      2017-03-25   Company       1718.0           0.00000

In [138]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.linear_model import SGDRegressor
from scipy.sparse import hstack
from sklearn.metrics import r2_score
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
import re
from pymorphy2 import MorphAnalyzer 

In [134]:
data["foreign_words"] = data.description.apply(lambda x: re.search(r"\b[a-zA-Z]+\b", x) is not None)

In [139]:
morph = MorphAnalyzer()

In [161]:
morph.parse("орел")[0].tag.POS

'NOUN'

In [192]:
class FasterMorphology:
    
    def __init__(self):
        self.__morph = MorphAnalyzer()
        self.__cache = {}
    
    def parse(self, word: str):
        if word in self.__cache:
            return self.__cache[word]
        else:
            analysis = self.__morph.parse(word)
            self.__cache[word] = analysis[0]
            return analysis[0]

In [ ]:
parser = FasterMorphology()
num_adjectives = np.zeros(shape)
num_nouns = np.zeros(shape)
num_verbs = np.zeros(shape)
num_adverbs = np.zeros(shape)
for idx in data.index:
    instance = data.loc[idx]
    shape = data.shape[0]
    for word in re.findall(r'(?u)\b\w\w+\b', instance.description):
        analysis = parser.parse(word)
        if analysis.tag.POS in ["ADJF", "ADJS"]:
            num_adjectives[idx] += 1
        elif analysis.tag.POS == "NOUN":
            num_nouns[idx] += 1
        elif analysis.tag.POS == "VERB":
            num_verbs[idx] += 1
        elif analysis.tag.POS == "ADVB":
            num_adverbs[idx] += 1
    if idx % 1000 == 0:
        print(idx)

In [202]:
data["num_verbs"] = num_verbs

In [203]:
data["num_adjectives"] = num_adjectives
data["num_adverbs"] = num_adverbs
data["num_nouns"] = num_nouns

In [204]:
data.tail()

parent_category_name               category_name          param_1  \
138710         Недвижимость           Земельные участки           Продам   
138711            Транспорт                  Автомобили       С пробегом   
138712      Для дома и дачи      Ремонт и строительство      Инструменты   
138713          Для бизнеса              Готовый бизнес         Торговля   
138714          Личные вещи  Товары для детей и игрушки  Детские коляски   

                param_2   param_3                             title  \
138710  Поселений (ИЖС)       NaN              Участок 7 сот. (ИЖС)   
138711       ВАЗ (LADA)  1111 Ока                ВАЗ 1111 Ока, 2004   
138712              NaN       NaN    Бензопила интерскол пцб 16\40л   
138713              NaN       NaN    Свадебный салон именно для вас   
138714              NaN       NaN  Продам прогулочную коляску Corol   

                                              description      price  \
138710  Продается земельный участок в поселке Совет-Кв...  1700000.0   
138711                                  Состояние хорошее    60000.0   
138712                           на запчасти,не заводится     1000.0   
138713  Продажа готового бизнеса в преддверии свадебно...   200000.0   
138714  Коляска легкая и крепкая, сезон зима-лето в хо...     4000.0   

        item_seq_number activation_date user_type  image_top_1  \
138710             1738      2017-03-23      Shop       1285.0   
138711                1      2017-03-15   Private       1055.0   
138712                9      2017-03-15   Private       1243.0   
138713               28      2017-03-18   Company        559.0   
138714                3      2017-03-20   Private        999.0   

        deal_probability  foreign_words  num_adjectives  num_verbs  \
138710           0.12205          False             3.0        1.0   
138711           0.73760          False             1.0        0.0   
138712           0.20123          False             0.0        1.0   
138713           0.00000          False             9.0        3.0   
138714           0.00000          False             3.0        1.0   

        num_adverbs  num_nouns  
138710          0.0       15.0  
138711          0.0        1.0  
138712          0.0        1.0  
138713          1.0       21.0  
138714          0.0       11.0

In [248]:
from sklearn.preprocessing import StandardScaler

In [251]:
scaler = StandardScaler()
data.num_adjectives = scaler.fit_transform(data.num_adjectives.values.reshape(-1, 1))

In [252]:
data.num_adverbs = scaler.fit_transform(data.num_adverbs.values.reshape(-1, 1))
data.num_nouns = scaler.fit_transform(data.num_nouns.values.reshape(-1, 1))
data.num_verbs = scaler.fit_transform(data.num_verbs.values.reshape(-1, 1))

In [241]:
def lemmatizer(text):
    return [parser.parse(word).normal_form for word in re.findall(r'(?u)\b\w\w+\b', text)]

In [245]:
vectorizer = TfidfVectorizer(ngram_range=(1, 3), min_df=3, tokenizer=lemmatizer)
X = vectorizer.fit_transform(data.description)

In [101]:
Y = data.deal_probability.copy()

In [206]:
X.shape

(138715, 67317)

In [253]:
gram_features = hstack([
            X,
            data.foreign_words.apply(lambda x: int(x)).values.reshape(data.shape[0],1), 
            data.num_adjectives.values.reshape(-1,1),
            data.num_verbs.values.reshape(-1,1),
            data.num_nouns.values.reshape(-1,1),
            data.num_adverbs.values.reshape(-1,1)
])

In [254]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, random_state=32)

In [258]:
model = SGDRegressor()
grid = GridSearchCV(model,
                    param_grid={"penalty": ["l1", "elasticnet"],
                                "alpha": [0.0001, 0.00001, 0.000001]},
                    scoring=["neg_mean_absolute_error"],
                    refit=False,
                    n_jobs=-1)

In [259]:
grid.fit(X_train, y_train)

/Users/grigoriyarshinov/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)


GridSearchCV(cv='warn', error_score='raise-deprecating',
             estimator=SGDRegressor(alpha=0.0001, average=False,
                                    early_stopping=False, epsilon=0.1,
                                    eta0=0.01, fit_intercept=True,
                                    l1_ratio=0.15, learning_rate='invscaling',
                                    loss='squared_loss', max_iter=1000,
                                    n_iter_no_change=5, penalty='l2',
                                    power_t=0.25, random_state=None,
                                    shuffle=True, tol=0.001,
                                    validation_fraction=0.1, verbose=0,
                                    warm_start=False),
             iid='warn', n_jobs=-1,
             param_grid={'alpha': [0.0001, 1e-05, 1e-06],
                         'penalty': ['l1', 'elasticnet']},
             pre_dispatch='2*n_jobs', refit=False, return_train_score=False,
             scoring=['neg_mean_

In [260]:
pd.DataFrame(grid.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_alpha  \
0       0.957252      0.005431         0.015288        0.001094      0.0001   
1       0.978234      0.008230         0.013093        0.002262      0.0001   
2       0.918702      0.045186         0.012341        0.002730       1e-05   
3       0.884003      0.014920         0.014244        0.000401       1e-05   
4       0.891446      0.005631         0.009974        0.002018       1e-06   
5       0.598221      0.200359         0.006849        0.000606       1e-06   

  param_penalty                                      params  \
0            l1          {'alpha': 0.0001, 'penalty': 'l1'}   
1    elasticnet  {'alpha': 0.0001, 'penalty': 'elasticnet'}   
2            l1           {'alpha': 1e-05, 'penalty': 'l1'}   
3    elasticnet   {'alpha': 1e-05, 'penalty': 'elasticnet'}   
4            l1           {'alpha': 1e-06, 'penalty': 'l1'}   
5    elasticnet   {'alpha': 1e-06, 'penalty': 'elasticnet'}   

   split0_test_neg_mean_absolute_error  split1_test_neg_mean_absolute_error  \
0                            -0.188470                            -0.187907   
1                            -0.187362                            -0.186680   
2                            -0.187210                            -0.186498   
3                            -0.186762                            -0.186062   
4                            -0.186731                            -0.186064   
5                            -0.186660                            -0.185951   

   split2_test_neg_mean_absolute_error  mean_test_neg_mean_absolute_error  \
0                            -0.189620                          -0.188666   
1                            -0.188318                          -0.187453   
2                            -0.188217                          -0.187308   
3                            -0.187768                          -0.186864   
4                            -0.187689                          -0.186828   
5                            -0.187657                          -0.186756   

   std_test_neg_mean_absolute_error  rank_test_neg_mean_absolute_error  
0                          0.000713                                  6  
1                          0.000672                                  5  
2                          0.000705                                  4  
3                          0.000700                                  3  
4                          0.000667                                  2  
5                          0.000700                                  1

Как дополнительные признаки я решил использовать наличие иностранных слов в тексте и количество слов определенных грамматических категорий. Первый может влиять на привлекательность товара как зарубежного, или попросту более подробно-описанного (Указание модели, фирмы-производителя). Морфологические признаки же позволяют обращать внимание модели на богатство используемых языковых ресурсов для описания товара.

С помощью GridSearchCV было выяснено, что лучший регрессор имеет параметры {'alpha': 1e-06, 'penalty': 'elasticnet'} и среднюю абсолютную ошибку **0.186756**